src/X11R6/contrib/programs/kinput2/cmd/kinput2.man
.\" $Id: kinput2.man,v 1.13 1994/06/06 06:50:22 ishisone Rel $
.TH KINPUT2 1 "6 June 1994" "X Version 11"
.SH NAME
kinput2 \- kanji input server for X11
.SH SYNOPSIS
\fBkinput2\fP [-options...]
.SH DESCRIPTION
.I Kinput2
is an X window application to input Japanese text.
It acts as a front-end for other applications that want
kana-kanji conversion.
.PP
When invoked, after some initialization (which will take about a few to
30 seconds depending on your machine) \fIkinput2\fP waits quietly for a
Japanese text input request from another client (i.e. no windows
appear).
When \fIkinput2\fP receives a request, it pops up a window and
starts conversion process. It sends the converted text back to the
client when the text is fixed.
.PP
.I Kinput2
has some big improvements over its predecessor, kinput, which
was contributed to X11R4.
.TP 8
\fBOver-the-spot Input\fP
This feature enables conversion process to be done at cursor position,
avoiding unnecessary movement of eyes between cursor position and conversion
window.
.I Kinput2
also supports root window style input and off-the-spot style input.
.TP 8
\fBMultiple Protocol Support\fP
.I Kinput2
supports several conversion protocols between front-end and
client. Supported protocols are:
.nf
	kinput protocol
	kinput2 protocol
	Matsushita jinput protocol
	Sony xlc protocol
	XIMP protocol
	X Input Method Protocol (X Consortium standard)
.fi
.TP 8
\fBMultiple Conversion Engine Support\fP
.I Kinput2
can use 3 different conversion engines, Wnn, Canna and Sj3.
You can choose one at compilation time, or at run time.
.TP 8
\fBKinput Compatibility\fP
.I Kinput2
is fully upward-compatible with kinput, so applications which use
kinput as the front-end can also use \fIkinput2\fP.
In this case, the applications cannot take advantage of over-the-spot
conversion, though.
.SH OPTIONS
.I kinput2
accepts all of the standard X Toolkit command line options.
The following options are accepted as well.
.TP 8
\fB\-wnn\fP
This option specifies that Wnn jserver is used as the conversion engine.
.TP 8
\fB\-jserver\fP
This option specifies the hostname of the jserver to be used for conversion.
.TP 8
\fB\-ccdef\fP
This option specifies the input character conversion rule file.
.TP 8
\fB\-canna\fP
This option specifies that Canna conversion server (irohaserver) is used as
the conversion engine.
.TP 8
\fB\-cannaserver\fP
This option specifies the hostname of Canna conversion server.
.TP 8
\fB\-cannafile\fP
This option specifies Canna conversion customization file.
.TP 8
\fB\-sj3\fP
This option specifies that Sj3 conversion server (sj3serv) is used as
the conversion engine.
.TP 8
\fB\-sj3serv\fP
This option specifies the hostname of the primary sj3 conversion server.
.TP 8
\fB\-sj3serv2\fP
This option specifies the hostname of the secondary sj3 server. The secondary
server is used when \fIkinput2\fP cannot connect to the primary server.
.TP 8
\fB\-font\fP \fIascii-font\fP
This option specifies the default font to be used for displaying ASCII text.
.TP 8
\fB\-kanjifont\fP \fIkanji-font\fP
This option specifies the default font to be used for displaying Kanji text.
.TP 8
\fB\-kanafont\fP \fIkana-font\fP
This option specifies the default font to be used for displaying Kana text.
.TP 8
\fB\-/+kinput\fP
This option enables/disables use of kinput protocol family (i.e.
kinput protocol, kinput2 protocol, jinput protocol and xlc protocol).
.TP 8
\fB\-/+ximp\fP
This option enables/disables use of XIMP protocol.
.TP 8
\fB\-/+xim\fP
This option enables/disables use of X Input Method Protocol.
.TP 8
\fB\-bc\fP
Kinput protocol, which is one of the communication protocols
\fIkinput2\fP supports,
uses X window's selection mechanism to notify other clients of
front-end's existence.
This option forces \fIkinput2\fP to use selection ``JAPANESE_CONVERSION''
(which is not a standard selection name) as well as ``_JAPANESE_CONVERSION''.
This is for backward compatibility, since many of the applications that
connect with kinput still use this non-standard selection.
.SH WIDGET TREE
Below is the widget hierarchy of \fIkinput2\fP.
.nf
.ta 0.5i 1.0i 1.5i 2.0i 2.5i 3.0i 3.5i 4.0i
	Kinput2  kinput2
		ConversionManager  convmanager
			KinputProtocol  kinputprotocol
			XimpProtocol  ximpprotocol
			IMProtocol  improtocol
			SeparateConversion  converter
				CcWnn  inputObj
				Canna  inputObj
				Sj3  inputObj
				JpWcharDisplay  displayObj
				Form  form
					ICLabel  mode
						JpWcharDisplay  display
					Canvas  text
				TransientShell  selectionShell
					CandidatePanel  selection
						JpWcharDisplay  display
				TransientShell  auxShell
					AuxPanel  aux
						JpWcharDisplay  display
			OffTheSpotConversion  converter
				CcWnn  inputObj
				Canna  inputObj
				Sj3  inputObj
				JpWcharDisplay  displayObj
				AdoptedShell  modeShell
					ICLabel  mode
						JpWcharDisplay display
				CanvasShell  text
				TransientShell  selectionShell
					CandidatePanel  selection
						JpWcharDisplay  display
				TransientShell  auxShell
					AuxPanel  aux
						JpWcharDisplay  display
			OverTheSpotConversion  converter
				CcWnn  inputObj
				Canna  inputObj
				Sj3  inputObj
				JpWcharDisplay  displayObj
				TransientShell  selectionShell
					CandidatePanel  selection
						JpWcharDisplay  display
				TransientShell  auxShell
					AuxPanel  aux
						JpWcharDisplay  display
				AdoptedShell  modeShell
					ICLabel  mode
						JpWcharDisplay  display
				TransientShell  modeShell
					ICLabel  mode
						JpWcharDisplay  display
				CanvasShell  text
.fi
.SH RESOURCES
There are too many resources to describe here, but you don't have to
know most of them.
So here is a brief description of resources which you might want to set.
.PP
\fBKinput2\fP application resource:
.RS 4
.TP 4
\fBconversionEngine\fP
This resource specifies which conversion engine to use.
You can specify one of ``wnn'', ``canna'' and ``sj3''.
.RE
.PP
.B ConversionControl
widget (which is a superclass
of SeparateConversion, OffTheSpotConversion and OverTheSpotConversion)
resource:
.RS 4
.TP 4
\fBsendbackKeyPress\fP
If this resource is ``true'', the KeyPress events intercepted but unused
by \fIkinput2\fP are sent back to the client when there is no conversion text.
Although this feature is convenient, it might confuse some clients
since none of corresponding KeyRelease events are sent.
Also, since the ``send_event'' flag of those events are true, some
clients do not accept events sent back by \fIkinput2\fP.
For example, kterm doesn't accept those events without specifying
.nf
	KTerm*allowSendEvents: true
.fi
in a resource file.
.br
The default of \fBsendbackKeyPress\fP resource is \fBtrue\fP.
.RE
.PP
.B OverTheSpotConversion
widget resources:
.RS 4
.TP 4
\fBshrinkWindow\fP
If this resource is ``true'', the width of on-the-spot conversion window
shrinks according to the length of the text on it. Otherwise,
the window never shrinks.
If the performance of Xserver on window resizing is poor,
you may consider setting this resources to false.
The default is \fBtrue\fP.
.TP 4
\fBmodeLocation\fP
This resource specifies the location of status widget in case of 
over-the-spot type conversion.
Unless client specifies the location explicitly, the location of
the status widget is determined by this resource.
If the value of this resource is ``topleft'', the widget is placed at the
top left of the client window. You can also specify 
``topright'', ``bottomleft'' and ``bottomright''.
If the value is ``tracktext'', the widget is placed just under the cursor
position, and moves according to the cursor. However, if the
region available for conversion is too small, status is placed at
the bottom left of the client area.
.br
The default is ``tracktext''.
.TP 4
\fBmodeBorderForeground\fP
If this resource is ``true'', the color of status widget is forced
to be same as the foreground color of the widget.
.br
The default is \fBfalse\fP.
.RE
.PP
.B CcWnn
(conversion widget using Wnn) resources:
.RS 4
.TP 4
\fBjserver\fP
This resource specifies the hostname of the (primary) jserver to be used
for conversion. If not specified, value of environment variable ``JSERVER''
is used.
.TP 4
\fBjserver2nd\fP
This resource specifies the hostname of the secondary jserver, which is 
used if the primary server is down. If not specified, \fIkinput2\fP doesn't
use secondary server.
.TP 4
\fBwnnEnvrc\fP
This resource specifies the filename of the Wnn conversion environment
initialization file. See documents of Wnn for the format of the file.
The default is \fB/usr/local/lib/wnn/wnnenvrc\fP.
.TP 4
\fBccdef\fP
This resource specifies the input character conversion rule file.
The default is \fB/usr/lib/X11/ccdef/ccdef.kinput2\fP.
.RE
.PP
.B Canna
(conversion widget using Canna) resources:
.RS 4
.TP 4
\fBcannahost\fP
This resource specifies the hostname of the canna server to be used.
.TP 4
\fBcannafile\fP
This resource specifies Canna conversion customization file.
.RE
.PP
.B Sj3
(conversion widget using SJ3) resources:
.RS 4
.TP 4
\fBsj3serv\fP
This resource specifies the hostname of the (primary) sj3serv to be used
for conversion. If not specified, value of environment variable ``SJ3SERV''
is used.
.TP 4
\fBsj3serv2\fP
This resource specifies the hostname of the secondary sj3serv, which is 
used if the primary server is down.
.TP 4
\fBrcfile\fP
This resource specifies the filename of the Sj3 conversion customization
file. The default is \fB/usr/lib/X11/sj3def/sjrc\fP.
.TP 4
\fBrkfile\fP
This resource specifies the Romaji-Kana conversion rule file.
The default is \fB/usr/lib/X11/sj3def/sjrk\fP.
.RE
.PP
.B JpWcharDisplay
(widget for drawing Japanese text) resources:
.RS 4
.TP 4
\fBfont\fP, \fBkanjiFont\fP, \fBkanaFont\fP
These resource specifies the font to be used for displaying ASCII,
Kanji and Kana text, respectively.
Since \fIkinput2\fP changes fonts dynamically by client requests,
the fonts specified by these resources might not be used.
If not specified, the following fonts are used.
.nf
	-*-fixed-medium-r-normal--14-*-*-*-c-*-iso8859-1
	-*-fixed-medium-r-normal--14-*-*-*-c-*-jisx0208.1983-0
	-*-fixed-medium-r-normal--14-*-*-*-c-*-jisx0201.1976-0
.fi
.RE
.PP
.B KinputProtocol
(widget for handling kinput protocol family \- kinput, kinput2,
jinput and xlc protocol) resources:
.RS 4
.TP 4
\fBbackwardCompatible\fP
All of the protocols in kinput protocol family
use X window's selection mechanism to notify other clients of
front-end's existence.
If this resource is ``true'', \fIkinput2\fP uses selection ``JAPANESE_CONVERSION''
(which is not a standard selection name) as well as ``_JAPANESE_CONVERSION''.
This is for backward compatibility, since many of the applications
using these protocols still use this non-standard selection.
.br
The default is \fBtrue\fP.
.TP 4
\fBxlcConversionStartKey\fP
This resource specifies the key that starts the conversion.
The key is a hint for the clients using xlc protocol to determine when
they should request conversion to \fIkinput2\fP.
The syntax of the specification is a subset of the translation table syntax:
.nf
	\fImodifier-names\fP<Key>\fIkeysym-name\fP
.fi
where modifier-names is a combination of \fBShift\fP, \fBLock\fP, \fBMeta\fP,
\fBAlt\fP and \fBMod[1-5]\fP.
.br
The default is \fBShift<Key>space\fP.
.RE
.PP
\fBXimpProtocol\fP and \fBIMProtocol\fP
(which handle XIMP protocol and X Input Method Protocol) resources:
.RS 4
.TP 4
\fBconversionStartKeys\fP
This resource specifies the keys that start the conversion.
The syntax of the specification is a subset of the translation table syntax:
.nf
	\fImodifier-names\fP<Key>\fIkeysym-name\fP
.fi
where modifier-names is a combination of \fBCtrl\fP, \fBShift\fP, \fBLock\fP,
\fBMeta\fP, \fBAlt\fP and \fBMod[1-5]\fP. A '~' can precede each modifier
name, meaning the modifier must not be asserted.
For example, if the conversion start key is ``Kanji'' key with shift key
pressed and control key not pressed (other modifier keys may be pressed or 
not), the specification is:
.nf
	Shift ~Ctrl<Key>Kanji
.fi
Multiple keys can be specified by separating them with newlines.
.br
The default is as follows:
.nf
	Shift<Key>space \en\e
	Ctrl<Key>Kanji
.fi
.TP 4
\fBdefaultFontList\fP
This resource specifies the default fonts to be used if the client
does not specify fonts to be used.
The default is \fB-*-fixed-medium-r-normal--14-*-*-*-c-*\fP.
.RE
.SH "CLIENT SETUP"
When using X Input Method Protocol, you have to specify the
name of the input server you want to use.
This can be done by setting ``\fBinputMethod\fP'' resource
or setting \fBXMODIFIERS\fP environment variable.
Since some clients don't regard inputMethod resource while other
clients ignore XMODIFIERS variable, so it is a good idea to set both of
them.
.PP
To use kinput2, you should add the following line to your
resource file (such as .Xresources or .Xdefaults):
.nf
	*inputMethod: kinput2
.fi
and, set XMODIFIERS to ``@im=kinput2''.
.nf
	setenv XMODIFIERS "@im=kinput2"\ \ \ (for csh)
	XMODIFIERS="@im=kinput2"; export XMODIFIERS\ \ \ (for sh)
.fi
.SH "SEE ALSO"
Wnn documents, Canna documents, Sj3 documents,
XIMP protocol specification,
The Input Method Protocol
.SH COPYRIGHT
.nf
Copyright 1988, 1989, 1990, 1991, 1992, 1994 Software Research Associates, Inc.
Copyright 1991 NEC Corporation, Tokyo, Japan.
Copyright 1991 Sony Corporation 
.fi
.SH AUTHORS
.nf
Makoto Ishisone, Software Research Associates, Inc.
Akira Kon, NEC Corporation
Naoshi Suzuki, Sony Corporation
.fi


src/NeWS
../NeWS

src/X11R6/contrib/programs/kinput2/cmd/kinput2.c
/*
 *	kinput2
 */

/*
 * Copyright (C) 1991 by Software Research Associates, Inc.
 *
 * Permission to use, copy, modify, and distribute this software and its
 * documentation for any purpose and without fee is hereby granted, provided
 * that the above copyright notice appear in all copies and that both that
 * copyright notice and this permission notice appear in supporting
 * documentation, and that the name of Software Research Associates not be
 * used in advertising or publicity pertaining to distribution of the
 * software without specific, written prior permission.  Software Research
 * Associates makes no representations about the suitability of this software
 * for any purpose.  It is provided "as is" without express or implied
 * warranty.
 *
 * Author:  Makoto Ishisone, Software Research Associates, Inc., Japan
 *		ishisone@sra.co.jp
 */

#ifndef lint
static char	*rcsid = "$Id: kinput2.c,v 1.34 1994/06/06 02:22:08 ishisone Rel $";
#endif

#include <stdio.h>
#include <signal.h>
#include <X11/Intrinsic.h>
#include <X11/StringDefs.h>
#include <X11/Shell.h>
#include "patchlevel.h"
#include "AsyncErr.h"
#include "MyDispatch.h"
#include "ConvMgr.h"
#include "KIProto.h"
#include "XimpProto.h"
#include "IMProto.h"

#if !defined(USE_WNN) && !defined(USE_CANNA) && !defined(USE_SJ3)
#define USE_WNN			/* default */
#endif

#ifdef USE_WNN
#include "CcWnn.h"
#include "WcharDisp.h"
#endif
#ifdef USE_CANNA
#include "Canna.h"
#include "WcharDisp.h"
#endif
#ifdef USE_SJ3
#include "Sj3.h"
#include "WcharDisp.h"
#endif

#include "DebugPrint.h"

int	debug_all;

/* application resource */
typedef struct {
    char *conversionEngine;
    int debugLevel;
    Boolean useKinputProtocol;
    Boolean useXimpProtocol;
    Boolean useXIMProtocol;
} AppResRec, *AppResP;

static AppResRec appres;

static XtResource app_resources[] = {
    { "conversionEngine", "ConversionEngine", XtRString, sizeof(String),
      XtOffset(AppResP, conversionEngine), XtRString, (XtPointer)"" },
    { "debugLevel", "DebugLevel", XtRInt, sizeof(int),
      XtOffset(AppResP, debugLevel), XtRImmediate, (XtPointer)0 },
    { "useKinputProtocol", "UseKinputProtocol", XtRBoolean, sizeof(Boolean),
      XtOffset(AppResP, useKinputProtocol), XtRImmediate, (XtPointer)True },
    { "useXimpProtocol", "UseXimpProtocol", XtRBoolean, sizeof(Boolean),
      XtOffset(AppResP, useXimpProtocol), XtRImmediate, (XtPointer)True },
    { "useXIMProtocol", "UseXIMProtocol", XtRBoolean, sizeof(Boolean),
      XtOffset(AppResP, useXIMProtocol), XtRImmediate, (XtPointer)True },
};

static XrmOptionDescRec	options[] = {
    {"-bc",		"*KinputProtocol.backwardCompatible", XrmoptionNoArg,	"True"},
    {"-font",		"*JpWcharDisplay.font",	XrmoptionSepArg,	NULL},
    {"-kanjifont",	"*JpWcharDisplay.kanjiFont", XrmoptionSepArg,	NULL},
    {"-kanafont",	"*JpWcharDisplay.kanaFont", XrmoptionSepArg,	NULL},
    {"-kinput",		".useKinputProtocol",	XrmoptionNoArg,		"on"},
    {"+kinput",		".useKinputProtocol",	XrmoptionNoArg,		"off"},
    {"-ximp",		".useXimpProtocol",	XrmoptionNoArg,		"on"},
    {"+ximp",		".useXimpProtocol",	XrmoptionNoArg,		"off"},
    {"-xim",		".useXIMProtocol",	XrmoptionNoArg,		"on"},
    {"+xim",		".useXIMProtocol",	XrmoptionNoArg,		"off"},
#ifdef DEBUG
    {"-debug",		".debugLevel",		XrmoptionNoArg,		"1"},
    {"-trace",		".debugLevel",		XrmoptionNoArg,		"10"},
    {"-debuglevel",	".debugLevel",		XrmoptionSepArg,	NULL},
#endif
#ifdef USE_WNN
    {"-wnn",		".conversionEngine",	XrmoptionNoArg,		"wnn"},
    {"-jserver",	"*CcWnn.jserver",	XrmoptionSepArg,	NULL},
    {"-ccdef",		"*CcWnn.ccdef",		XrmoptionSepArg,	NULL},
    {"-wnnenvname",	"*CcWnn.wnnEnvname",	XrmoptionSepArg,	NULL},
    {"-wnnenvrc",	"*CcWnn.wnnEnvrc",	XrmoptionSepArg,	NULL},
#endif
#ifdef USE_CANNA
    {"-canna",		".conversionEngine",	XrmoptionNoArg,		"canna"},
    {"-cannaserver",	"*Canna.cannahost",	XrmoptionSepArg,	NULL},
    {"-cs",		"*Canna.cannahost",	XrmoptionSepArg,	NULL},
    {"-cannafile",	"*Canna.cannafile",	XrmoptionSepArg,	NULL},
#endif
#ifdef USE_SJ3
    {"-sj3",        ".conversionEngine",    XrmoptionNoArg,     "sj3"},
    {"-sj3serv",    "*Sj3.sj3serv",     XrmoptionSepArg,    NULL},
    {"-sj3serv2",   "*Sj3.sj3serv2",    XrmoptionSepArg,    NULL},
    {"-sj3user",    "*Sj3.sj3user",     XrmoptionSepArg,    NULL},
    {"-rcfile",     "*Sj3.rcfile",      XrmoptionSepArg,    NULL},
    {"-sbfile",     "*Sj3.sbfile",      XrmoptionSepArg,    NULL},
    {"-rkfile",     "*Sj3.rkfile",      XrmoptionSepArg,    NULL},
    {"-hkfile",     "*Sj3.hkfile",  XrmoptionSepArg,    NULL},
    {"-zhfile",     "*Sj3.zhfile",  XrmoptionSepArg,    NULL},
    {"-sjrc",       "*Sj3.rcfile",      XrmoptionSepArg,    NULL},
    {"-sjsb",       "*Sj3.sbfile",      XrmoptionSepArg,    NULL},
    {"-sjrk",       "*Sj3.rkfile",      XrmoptionSepArg,    NULL},
    {"-sjhk",       "*Sj3.hkfile",      XrmoptionSepArg,    NULL},
    {"-sjzh",       "*Sj3.zhfile",      XrmoptionSepArg,    NULL},
#endif
};

XtAppContext	apc;
Widget		toplevel;

static int	numProtocols;
static int	(*DefaultErrorHandler)();

static WidgetClass	getInputObjClass();
static int	IgnoreBadWindow();
#ifdef SIGNALRETURNSINT
static int	scheduleExit();
#else
static void	scheduleExit();
#endif
static void	exitTimer();
static void	Destroyed();
static void	Exit();
static void	realExit();
static void	usage();
static void	print_version();

void
main(ac, av)
int ac;
char **av;
{
    Widget manager, protocol;
    int i;
    WidgetClass inputobjclass, displayobjclass;

    toplevel = XtAppInitialize(&apc, "Kinput2",
			       options, XtNumber(options),
			       &ac, av,
			       (String *)NULL, (ArgList)NULL, 0);

    /* check invalid argument */ 
    if (ac > 1) {
	int do_usage = 0;
	for (i = 1; i < ac; i++) {
	    if (!strcmp(av[i], "-version")) {
		print_version();
	    } else {
		fprintf(stderr, "unknown argument: %s\n", av[i]);
		do_usage = 1;
	    }
	}
	if (do_usage) usage();
    }

    /* initialize asynchronous error handler */
    XAEInit();

    XtGetApplicationResources(toplevel, &appres,
			      app_resources, XtNumber(app_resources),
			      NULL, 0);

    /* set debug level */
    debug_all = appres.debugLevel;

#ifdef RANDOM_ID
    /*
     * one nasty hack here:
     *
     * kinput clients often use server's window ID for the only key
     * value to identify their conversion server (kinput), and they
     * think it is dead and take appropriate action (eg connecting to
     * the new server) when they notice the ID has changed.
     *
     * but it is likely that another kinput has the same resource ID
     * base (because X servers always choose the smallest unused ID
     * base for new clients). and if it is the same, so is the owner
     * window ID, and the clients don't notice the change.
     *
     * to get rid of the problem, we add some small random offset to
     * the resource ID so that every time we get different owner ID
     * even if the resource ID base is the same.
     *
     * of course it heavily depends on the current implementaion of
     * the resource ID allocation in Xlib, so I call it 'nasty'.
     */
    XtDisplay(toplevel)->resource_id += getpid() % 1024;
#endif

    inputobjclass = getInputObjClass();
    displayobjclass = jpWcharDisplayObjectClass;

    manager = XtVaCreateManagedWidget("convmanager",
				      conversionManagerWidgetClass,
				      toplevel,
				      XtNwidth, 1,
				      XtNheight, 1,
				      NULL);

    numProtocols = 0;

    if (appres.useKinputProtocol) {
	protocol = XtVaCreateWidget("kinputprotocol",
				    kinputProtocolWidgetClass,
				    manager,
				    XtNlanguage, "JAPANESE",
				    XtNinputObjectClass, inputobjclass,
				    XtNdisplayObjectClass, displayobjclass,
				    XtNwidth, 1,
				    XtNheight, 1,
				    NULL);
	XtAddCallback(protocol, XtNdestroyCallback,
		      Destroyed, (XtPointer)NULL);
	numProtocols++;
    }

    if (appres.useXimpProtocol) {
	protocol = XtVaCreateWidget("ximpprotocol",
				    ximpProtocolWidgetClass,
				    manager,
				    XtNlocaleName, "ja_JP",
				    XtNinputObjectClass, inputobjclass,
				    XtNdisplayObjectClass, displayobjclass,
				    XtNwidth, 1,
				    XtNheight, 1,
				    NULL);
	XtAddCallback(protocol, XtNdestroyCallback,
		      Destroyed, (XtPointer)NULL);
	numProtocols++;
    }

    if (appres.useXIMProtocol) {
	protocol = XtVaCreateWidget("improtocol",
				    imProtocolWidgetClass,
				    manager,
				    XtNlanguage, "ja_JP",
				    XtNinputObjectClass, inputobjclass,
				    XtNdisplayObjectClass, displayobjclass,
				    XtNwidth, 1,
				    XtNheight, 1,
				    NULL);
	XtAddCallback(protocol, XtNdestroyCallback,
		      Destroyed, (XtPointer)NULL);
	numProtocols++;
    }

    if (numProtocols == 0) {
	fprintf(stderr, "no protocols activated\n");
	exit(1);
    }

    /* set signal handler */
    if (signal(SIGINT, SIG_IGN) != SIG_IGN) signal(SIGINT, scheduleExit);
    signal(SIGTERM, scheduleExit);
#ifdef USE_WNN
    signal(SIGPIPE, SIG_IGN);
#endif

    /* set my error handler */
    DefaultErrorHandler = XAESetErrorHandler(IgnoreBadWindow);

    XtRealizeWidget(toplevel);

    for (;;) {
	XEvent event;

	XtAppNextEvent(apc, &event);
	XtDispatchEvent(&event);
	MyDispatchEvent(&event); /* additional dispatcher */
    }
    /* NOTREACHED */
}

static WidgetClass
getInputObjClass()
{
    WidgetClass class;

#ifdef USE_WNN
    if (!strcmp(appres.conversionEngine, "wnn")) {
	return ccWnnObjectClass;
    }
#endif
#ifdef USE_CANNA
    if (!strcmp(appres.conversionEngine, "canna") ||
	!strcmp(appres.conversionEngine, "iroha")) {
	return cannaObjectClass;
    }
#endif
#ifdef USE_SJ3
    if (!strcmp(appres.conversionEngine, "sj3")) {
	return sj3ObjectClass;
    }
#endif

    /* set default input object */
#ifdef USE_SJ3
    class = sj3ObjectClass;
#endif
#ifdef USE_CANNA
    class = cannaObjectClass;
#endif
#ifdef USE_WNN
    class = ccWnnObjectClass;
#endif

    return class;
}

static int
IgnoreBadWindow(dpy, error)
Display *dpy;
XErrorEvent *error;
{
    /*
     * BadWindow events will be sent if any of the active clients die
     * during conversion.  Although I select DestroyNotify event on the
     * client windows to detect their destruction and take appropriate
     * actions, and I'm careful when doing critical operations, but still
     * there is a chance of getting unexpecting BadWindow error caused by
     * client death.
     * There are also chances of getting BadDrawable as well.
     * So I set the error handler to ignore BadWindow/BadDrawable errors.
     * Of course I'd better check if the resourceid field of the error
     * event is the window ID of a client, but I'm too lazy to do that...
     */
    if (error->error_code != BadWindow && error->error_code != BadDrawable) {
	/* invoke default error handler */
	(*DefaultErrorHandler)(dpy, error);
    }
    return 0;
}

#ifdef SIGNALRETURNSINT
static int
#else
static void
#endif
scheduleExit()
{
    /*
     * It is unwise to do complex operation (in this case,
     * XtDestroyWidget) in a signal handler.
     * So postpone the real work...
     */
    XtAppAddTimeOut(apc, 1L, exitTimer, (XtPointer)NULL);
}

/* ARGSUSED */
static void
exitTimer(cldata, timerp)
XtPointer cldata;
XtIntervalId *timerp;
{
    Exit();
}

/* ARGSUSED */
static void
Destroyed(w, client_data, call_data)
Widget w;
XtPointer client_data;
XtPointer call_data;
{
    static int cnt = 0;

    /*
     * if all the protocol widgets are dead, kill myself.
     */
    if (++cnt >= numProtocols) Exit();
}

static void
Exit()
{
    static int exiting = 0;

    if (exiting) return;

    exiting = 1;

    /*
     * Destroy all widgets.
     */
    XtDestroyWidget(toplevel);

    /*
     * Postpone calling exit() until next XtNextEvent(),
     * in order to give each widget time to execute their
     * destroy procedure.
     */
    XtAppAddTimeOut(apc, 1L, realExit, (XtPointer)NULL);
}

/* ARGSUSED */
static void
realExit(cldata, timerp)
XtPointer cldata;
XtIntervalId *timerp;
{
    exit(0);
}

static void
usage()
{
    char **p;
    static char *syntaxtable[] = {
#ifdef USE_WNN
	"-wnn",			"use Wnn as the conversion engine",
	"-jserver <hostname>",	"specify jserver host",
	"-ccdef <ccdeffile>",	"specify character conversion def. file",
	"-wnnenvname <name>",	"specify Wnn environment name",
	"-wnnenvrc <wnnenvrcfile>",	"specify Wnn environment file",
#endif
#ifdef USE_CANNA
	"-canna",		"use Canna (Iroha) as the conversion engine",
	"{-cannaserver|-cs} <hostname>[:n]", "specify cannaserver host",
	"-cannafile <cannafile>", "specify canna customize file",
#endif
#ifdef USE_SJ3
	"-sj3",         	"use SJ3 as the conversion engine",
	"-sj3serv <hostname>",  "specify first sj3serv host",
	"-sj3serv2 <hostname>", "specify second sj3serv host",
	"-sj3user <user>",      "specify user name connect to sj3serv",
	"{-rcfile|-sjrc} <file>",       "specify resource definition file",
	"{-sbfile|-sjsb} <file>",       "specify symbol table file",
	"{-rkfile|-sjrk} <file>",       "specify roma-kana coversion definition file",
	"{-hkfile|-sjhk} <file>",       "specify hira-kata coversion definition file",
	"{-zhfile|-sjzh} <file>",       "specify zen/han coversion definition file",
#endif
	"-bc",			"backward compatible mode",
	"-font <font>",		"ASCII font to be used",
	"-kanjifont <font>",	"KANJI font to be used",
	"-kanafont <font>",	"KANA font to be used",
	"-/+kinput",		"activate/deactivate kinput protocol family",
	"-/+ximp",		"activate/deactivate Ximp protocol",
	"-/+xim",		"activate/deactivate X Input Method protocol",
	"-background <color>",	"background color",
	"-foreground <color>",	"foreground color",
	"-rv",			"reverse video mode",
	"-display <display>",	"specify display",
	"-version",		"print version information and exit",
#ifdef DEBUG
	"-debug",		"print debug messages (debug level 1)",
	"-trace",		"print trace messages (debug level 10)",
	"-debuglevel <level>",	"set debug level",
#endif
	NULL, NULL,
    };

    fprintf(stderr, "options are:\n");
    for (p = syntaxtable; *p != NULL; p += 2) {
	fprintf(stderr, "    %-30s %s\n", *p, *(p + 1));
    }
    exit(1);
}

static void
print_version()
{
    char *p;

    printf("kinput2 %s ", KINPUT2_VERSION);
    if (PATCHLEVEL > 0) printf("fix %d ", PATCHLEVEL);
    printf(" (");
    p = DATE + 7;				/* skip '$Date: ' */
    while (*p != '\0' && *p != ' ') {
	putchar(*p);	/* print date */
	p++;
    }
    printf(")\n");

    printf("\toptions: ");
#ifdef USE_WNN
    printf("[Wnn] ");
#endif
#ifdef USE_CANNA
    printf("[Canna2] ");
#endif
#ifdef USE_SJ3
    printf("[Sj3] ");
#endif
#ifdef DEBUG
    printf("[DEBUG] ");
#endif
    printf("\n");
    exit(0);
}

#if defined(USE_WNN) && defined(NEED_Strlen)
/*
 * Wnn/jlib/js.c should have this function...
 */
int
Strlen(s)
unsigned short *s;
{
    int n = 0;

    while (*s++) n++;
    return n;
}
#endif
